sources/imdsclient/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+/*!
+The imdsclient library provides high-level methods to interact with the AWS Instance Metadata Service.
+The high-level methods provided are [`fetch_dynamic`], [`fetch_metadata`], and [`fetch_userdata`].
+
+For more control, and to query IMDS without high-level wrappers, there is also a [`fetch_imds`] method.
+This method is useful for specifying things like a pinned date for the IMDS schema version.
+*/
+
+#![deny(rust_2018_idioms)]
+
+use http::StatusCode;
+use log::{debug, info, trace, warn};
+use reqwest::blocking::Client;
+use snafu::{ensure, ResultExt};
+
+const IMDS_BASE_URI: &str = "http://169.254.169.254";
+const IMDS_PINNED_DATE: &str = "2020-10-27";
+
+// Currently only able to get fetch session tokens from `latest`
+const IMDS_SESSION_TOKEN_URI: &str = "http://169.254.169.254/latest/api/token";
+
+/// A client for making IMDSv2 queries.
+/// It obtains a session token when it is first instantiated and is reused between helper functions.
+pub struct ImdsClient {
+    client: Client,
+    session_token: String,
+}
+
+impl ImdsClient {
+    pub fn new() -> Result<Self> {
+        let client = Client::new();
+        let session_token = fetch_token(&client)?;
+        Ok(Self {
+            client,
+            session_token,
+        })
+    }
+
+    /// Helper to fetch `dynamic` targets from IMDS, preferring an override file if present.
+    pub fn fetch_dynamic(&mut self, end_target: &str, description: &str) -> Result<Option<String>> {
+        let dynamic_target = format!("dynamic/{}", end_target);
+        self.fetch_imds(IMDS_PINNED_DATE, &dynamic_target, description)
+    }
+
+    /// Helper to fetch `meta-data` targets from IMDS, preferring an override file if present.
+    pub fn fetch_metadata(
```

Comment:
I'm a bit worried about this API since we're not handling these strings as URLs and we're not doing any handling of slashes, etc.  Ideally we'd at least try to parse the string as a URL so we at least know it's valid.

Reply:
I was going to rely on the `str` -> `Url` parse in `Reqwest`'s `get` function, but it probably makes sense to parse the `uri`  just before the loop.